ui/StatusQ/src/fastexpressionrole.cpp

```diff
@@ -0,0 +1,164 @@
+    allowing to implement a custom role based on a javascript expression.
+    However in FastExpressionRole's expression context there are available only
+    roles explicitly listed in \l expectedRoles property:
+
+    \code
+    SortFilterProxyModel {
+       sourceModel: numberModel
+       proxyRoles: FastExpressionRole {
+           name: "c"
+           expression: model.a + model.b
+           expectedRoles: ["a", "b"]
+      }
+    }
+    \endcode
+
+    By accessing only needed roles, the performance is significantly better in
+    comparison to ExpressionRole, especially when the model has multiple
+    FastExpressionRole's.
+*/
+
+/*!
+    \qmlproperty expression FastExpressionRole::expression
+
+    See ExpressionRole::expression for details. Unline the original
+    ExpressionRole only roles explicitly declared via expectedRoles are accessible.
+*/
+const QQmlScriptString& FastExpressionRole::expression() const
+{
+    return m_scriptString;
+}
+
+void FastExpressionRole::setExpression(const QQmlScriptString& scriptString)
+{
+    if (m_scriptString == scriptString)
+        return;
+
+    m_scriptString = scriptString;
+    updateExpression();
+
+    emit expressionChanged();
+    invalidate();
+}
+
+void FastExpressionRole::proxyModelCompleted(const QQmlSortFilterProxyModel& proxyModel)
+{
+    updateContext(proxyModel);
+}
+
+void FastExpressionRole::setExpectedRoles(const QStringList& expectedRoles)
+{
+    if (m_expectedRoles == expectedRoles)
+        return;
+
+    m_expectedRoles = expectedRoles;
+    emit expectedRolesChanged();
+
+    invalidate();
+}
+
+/*!
+    \qmlproperty list<string> FastExpressionRole::expectedRoles
+
+    List of role names intended to be available in the expression's context.
+*/
+const QStringList& FastExpressionRole::expectedRoles() const
+{
+    return m_expectedRoles;
+}
+
+QVariant FastExpressionRole::data(const QModelIndex& sourceIndex,
+                                  const QQmlSortFilterProxyModel& proxyModel)
+{
+    if (m_scriptString.isEmpty())
+        return {};
+
+    QVariantMap modelMap;
+    QHash<int, QByteArray> roles = proxyModel.roleNames();
+
+    QQmlContext context(qmlContext(this));
+    auto addToContext = [&] (const QString &name, const QVariant& value) {
+        context.setContextProperty(name, value);
+        modelMap.insert(name, value);
+    };
+
+    for (auto it = roles.cbegin(); it != roles.cend(); ++it) {
+        auto name = it.value();
+
+        if (!m_expectedRoles.contains(name))
+            continue;
+
+        addToContext(name, proxyModel.sourceData(sourceIndex, it.key()));
+    }
+
+    addToContext(QStringLiteral("index"), sourceIndex.row());
+
+    context.setContextProperty(QStringLiteral("model"), modelMap);
+
+    QQmlExpression expression(m_scriptString, &context);
+    QVariant result = expression.evaluate();
+
+    if (expression.hasError())
+        qWarning() << expression.error();
```

Comment:
You should `return {}` here, not the `result` below

Reply:
In case of error `expression.evaluate()` returns invalid QVariant, same as the one constructed by `{}`. So it's the same but simpler.